File: src/main/java/komu/blunt/asm/opcodes/OpStoreVariable.java
package komu.blunt.asm.opcodes;

import komu.blunt.analyzer.VariableReference;
import komu.blunt.asm.Register;
import komu.blunt.asm.VM;
import komu.blunt.eval.Environment;

import static com.google.common.base.Preconditions.checkNotNull;

public final class OpStoreVariable extends OpCode {
    private final VariableReference variable;
    private final Register register;

    public OpStoreVariable(VariableReference variable, Register register) {
        this.variable = checkNotNull(variable);
        this.register = checkNotNull(register);
    }

    @Override
    public void execute(VM vm) {
        Environment env = variable.isGlobal() ? vm.getGlobalEnvironment() : (Environment) vm.get(Register.ENV);
        Object value = vm.get(register);
        env.set(variable, value);
    }

    @Override
    public String toString() {
        return String.format("(store (variable %d %d) %s) ; %s", variable.frame, variable.offset, register, variable.name);
    }
}

File: src/main/kotlin/komu/blunt/asm/opcodes/OpStoreVariable.kt
package komu.blunt.asm.opcodes

import komu.blunt.analyzer.VariableReference
import komu.blunt.asm.Register
import komu.blunt.asm.VM

class OpStoreVariable(private val variable: VariableReference, private val register: Register) : OpCode() {

    override fun execute(vm: VM) {
        val env = if (variable.global) vm.globalEnvironment else vm.env
        env[variable] = vm[register]
    }

    override fun modifies(register: Register) = false
    override fun toString() = "(store (variable ${variable.frame} ${variable.offset}) $register) ; ${variable.name}"
}
